### PR TITLE
internal cmd: GHE repo feeder

### PR DESCRIPTION
**Pull request by uwedeportivo**
internal cmd tool that feeds a GHE instance with public repos from github.com.

used to replicate customer setups and for stress testing an instance

### CODE REVIEW

File: internal/cmd/ghe-feeder/main.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+	"io/ioutil"
+	"log"
+	"math"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/briandowns/spinner"
```
Comment (keegancsmith) [inline internal/cmd/ghe-feeder/main.go:16]: we recently introduced a similiar dependency, except maintained/written by @efritz :) Maybe use that instead? https://github.com/efritz/pentimento

Comment (uwedeportivo) [reply internal/cmd/ghe-feeder/main.go:16]: sure, no worries

**Comment by uwedeportivo** [comment]
ready for review

File: internal/cmd/ghe-feeder/namesgenerator.go [hunk 2]
```diff
@@ -0,0 +1,852 @@
...
+
+		// Nikolay Yegorovich Zhukovsky (Russian: Никола́й Его́рович Жуко́вский, January 17 1847 – March 17, 1921) was a Russian scientist, mathematician and engineer, and a founding father of modern aero- and hydrodynamics. Whereas contemporary scientists scoffed at the idea of human flight, Zhukovsky was the first to undertake the study of airflow. He is often called the Father of Russian Aviation. https://en.wikipedia.org/wiki/Nikolay_Yegorovich_Zhukovsky
+		"zhukovsky",
+	}
+)
+
+// GetRandomName generates a random name from the list of adjectives and surnames in this package
+// formatted as "adjective_surname". For example 'focused_turing'. If retry is non-zero, a random
+// integer between 0 and 10 will be added to the end of the name, e.g `focused_turing3`
+func getRandomName(retry int) string {
```
Comment (bobheadxi) [inline internal/cmd/ghe-feeder/namesgenerator.go:841]: is this sourced from Docker? I think there are third-party libraries that can do something similar without including all this in the repo, ie https://github.com/Pallinder/go-randomdata

File: go.mod [hunk 3]
```diff
@@ -44,6 +44,7 @@ require (
 	github.com/google/go-cmp v0.4.0
 	github.com/google/go-github v17.0.0+incompatible
 	github.com/google/go-github/v28 v28.1.1
+	github.com/google/go-github/v31 v31.0.0
```
Comment (unknwon) [inline go.mod:47]: This is when we want Go modules, we imported all major versions of `github.com/google/go-github` in one repository 😄

File: internal/cmd/ghe-feeder/README.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+## Inputs
+
+Inputs are command line arguments consisting of files and directories. Files need to end in .json, .txt or .csv.
+Directories are traversed recursively for such files.
+
+If the input file is a .json or .txt file it expects an owner/repo string per line (and nothing else). For example:
+
+```
+"FrankBGao/try"
```
Comment (unknwon) [inline internal/cmd/ghe-feeder/README.md:60]: Are double quotes required? Seems weird in a .txt file.

File: internal/cmd/ghe-feeder/ghe-feeder-dashboard.json [hunk 5]
```diff
@@ -0,0 +1,625 @@
+{
```
Comment (unknwon) [inline internal/cmd/ghe-feeder/ghe-feeder-dashboard.json:1]: Can you describe in README what does this file do?

File: internal/cmd/ghe-feeder/main.go [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+	if len(*scratchDir) == 0 {
+		d, err := ioutil.TempDir("", "ghe-feeder")
+		if err != nil {
+			log15.Error("failed to create scratch dir", "error", err)
+			os.Exit(1)
+		}
+		*scratchDir = d
+	}
+
+	u, err := urlx.Parse(*baseURL)
```
Comment (unknwon) [inline internal/cmd/ghe-feeder/main.go:95]: What's good about `urlx`?

File: internal/cmd/ghe-feeder/progress.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+		_, err = stmt.Exec(ownerRepo)
+		if err != nil {
+			return false, err
+		}
+
+		return false, nil
+	}
+
+	alreadyDone = !failed || (failed && errType == "clone")
+	return
```
Comment (unknwon) [inline internal/cmd/ghe-feeder/progress.go:84]:
```suggestion
	return alreadyDone, nil
```

File: internal/cmd/ghe-feeder/namesgenerator.go (hunk 2 above)
Comment (uwedeportivo) [reply internal/cmd/ghe-feeder/namesgenerator.go:841]: ahhh. i'm too deep into this scheme now to change (1k orgs generated already). i prefer to keep this.

File: internal/cmd/ghe-feeder/README.md (hunk 4 above)
Comment (uwedeportivo) [reply internal/cmd/ghe-feeder/README.md:60]: they're not required. i adjusted the code to use whatever we found as input

File: internal/cmd/ghe-feeder/ghe-feeder-dashboard.json (hunk 5 above)
Comment (uwedeportivo) [reply internal/cmd/ghe-feeder/ghe-feeder-dashboard.json:1]: sure

File: go.mod (hunk 3 above)
Comment (uwedeportivo) [reply go.mod:47]: :-)

File: internal/cmd/ghe-feeder/main.go (hunk 6 above)
Comment (uwedeportivo) [reply internal/cmd/ghe-feeder/main.go:95]: oh, didn't realize Goland picked that one. will fix

**Comment by uwedeportivo** [comment]
script works well. it's been used for days, has very low mem, pumps repos into GHE like crazy and behaves well. 140k repos added and 1k+ orgs. is very flexible with resume points and skipping and progress. dashboard in grafana is super helpful to see what's going on

**Pull request merged**